python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+
+        if resolve:
+            await self.commit()
+
+        try:
+            return self.get(key, default)
+        finally:
+            await self.commit()
+
+    def lock(self):
+        """
+        Lock the cache to prevent other threads or processes from editing data
+
+        Does nothing by default. Override in subclass to add locking capability.
+        """
+
+    def release(self):
+        """
+        Unlock the cache to allow other threads or processes to edit data
+
+        Does nothing by default. Override in subclass to add locking capability.
+        """
+
+    def is_locked(self) -> bool:
+        """
+        Whether the cache is currently locked
+
+        Does nothing by default. Override in subclass to add locking capability.
+        """
+        return False
+
+    def update(
+        self,
+        other: typing.Union[
+            InputCatalog[HashableTypeT],
+            typing.Mapping[str, typing.Union[CatalogEntry[HashableTypeT], HashableTypeT, None]]
+        ],
+        **kwargs
+    ) -> None:
+        """
+        Merge data into this cache
+
+        Triggered asynchronous tasks will be stored for later completion
+
+        Args:
+            other: Another cache or mapping of data to merge into this cache
+            kwargs: Keyvalue pairs of data to add to this cache
+        """
+        try:
+            self.lock()
+            if isinstance(other, InputCatalog):
+                for entry in other.values():
+                    self[entry.identifier] = CatalogEntry(entry.identifier, entry)
+            elif isinstance(other, typing.Mapping):
+                for key, value in other.items():
+                    self[key] = CatalogEntry(key, value)
+            else:
+                raise ValueError(f"Cannot update a {self.__class__.__name__} with a value of type '{type(other)}'")
+
+            for key, value in kwargs.items():
+                self[key] = CatalogEntry(key, value)
+        finally:
+            self.__update_earliest()
+            self.release()
+
+    def __order_entries(self, descending: bool = None) -> typing.Sequence[HashableTypeT]:
+        """
+        Create a collection of all cached data in access time order
+
+        Args:
+            descending: Whether to order entries with the most recently accessed items first
+
+        Returns:
+            A sequence of all entries in access time order
+        """
+        try:
+            self.lock()
+
+            if descending is None:
+                descending = True
+
+            sorted_wrappers = sorted(self.values(), key=lambda entry: entry.last_accessed, reverse=descending)
+            sorted_values = [
+                wrapper.data
+                for wrapper in sorted_wrappers
+            ]
+            return sorted_values
+        finally:
+            self.release()
+
+    def __iter__(self) -> typing.Iterator[CatalogEntry[HashableTypeT]]:
+        return iter(self.__order_entries())
```

Comment:
Needs to be noted that items will not change from under you while iterating, but there are no guarantees that the items are fresh or even still present.